src/bluez/adapter/peripheral.rs

```diff
@@ -11,410 +11,404 @@
+            let services_resolved = services_resolved.0.as_u64().unwrap() > 0;
+            if services_resolved {
+                // Need to prase and figure out handle ranges for all discovered characteristics.
+                self.build_characteristic_ranges().unwrap();
             }
-            msg => {
-                debug!("ignored message {:?}", msg);
+            // All services have been discovered, time to inform anyone waiting.
+            let (ref lock, ref cvar) = *self.state;
+            if services_resolved {
+                *lock.lock().unwrap() = PeripheralState::ServicesResolved;
             }
+            cvar.notify_all();
         }
-    }
 
-    fn request_raw_async(&self, data: &mut [u8], handler: Option<RequestCallback>) {
-        let l = self.stream.read().unwrap();
-        match l.as_ref().ok_or(Error::NotConnected) {
-            Ok(stream) => {
-                stream.write(&mut *data, handler);
-            }
-            Err(err) => {
-                if let Some(h) = handler {
-                    h(Err(err));
+        // if let Some(services) = args.get("ServiceData") {
+        //     debug!("Updating services to \"{:?}\"", services);
+
+        //     if let Some(mut iter) = services.0.as_iter() {
+        //         loop {
+        //             if let (Some(uuid), ())
+        //         }
+        //     }
+        // }
+
+        // As of writing this: ManufacturerData returns a 'Variant({<manufacturer_id>: Variant([<manufacturer_data>])})'.
+        // This Variant wrapped dictionary and array is difficult to navigate. So uh.. trust me, this works on my machine™.
+        if let Some(manufacturer_data) = args.get("ManufacturerData") {
+            debug!(
+                "Updating \"{}\" manufacturer data \"{:?}\"",
+                self.address, manufacturer_data
+            );
+            let mut result = Vec::<u8>::new();
+            // dbus-rs doesn't really have a dictionary API... so need to iterate two at a time and make a key-value pair.
+            if let Some(mut iter) = manufacturer_data.0.as_iter() {
+                loop {
+                    if let (Some(id), Some(data)) = (iter.next(), iter.next()) {
+                        // This API is terrible.. why can't I just get an array out, why is it all wrapped in a Variant?
+                        let data: Vec<u8> = data
+                            .as_iter() // 🎶 The Variant is connected to the
+                            .unwrap() // Array type!
+                            .next() // The Array type is connected to the
+                            .unwrap() // Array of integers!
+                            .as_iter() // Lets convert the
+                            .unwrap() // integers to a
+                            .map(|b| b.as_u64().unwrap() as u8) // array of bytes...
+                            .collect(); // I got too lazy to make it rhyme... 🎶
```

Comment:
You can do this more nicely with `dbus::arg::prop_cast`, or if you pass `--prop-newtype` to `dbus-codegen-rust` it will give you a type-safe wrapper for the `HashMap`, and you can do `OrgBluezDevice1Properties::manufacturer_data` to get a `HashMap<u16, Variant<Box<dyn RefArg>>>`, and then you can just map the values with `dbus::arg::cast::<Vec<u8>>(&v.0)`.

Reply:
From memory, I gave this a try from the very start. but I think because i'm passing in a subset of the type that `dbus::arg::cast/prop_cast` are expecting, i get compile errors. I'll have another go and see if I've missed something.

Reply:
Ah, you may also need to make sure you have at least version 0.9.1 of the `dbus` crate, to include the fix from https://github.com/diwic/dbus-rs/pull/306.

Reply:
I've sent you https://github.com/NZSmartie/btleplug/pull/1 to fix this.

Reply:
Thanks @qwandor! Though i'm a bit sad to see my silly rhyming comment go away 😢

Reply:
We need to keep some standards in this code, so I'm at least going to need some methods in iambic pentameter going forward.